### PR TITLE
Fix tnp.logaddexp gradients when both arguments are equal

### PR DESCRIPTION
**Pull request by Phoebus-Liu**
## 关联 issue

Related: #127232

The issue reproduces on the `tf.keras.ops.numpy.logaddexp` path, which is owned by the Keras package and is not touched here. This PR fixes the sibling implementation that lives in this repository, `tf.experimental.numpy.logaddexp`, which has the same defect.

## 问题

`tf.experimental.numpy.logaddexp` is implemented as

```python
maximum(x1, x2) + log1p(exp(-abs(x1 - x2)))
```

`maximum` sends the whole gradient to its first argument where the two are equal, so at `x1 == x2` the partial derivatives evaluate to `1.0` and `0.0` instead of the analytic `0.5` and `0.5`; the second derivatives are `0.0` instead of `0.25`.

## 改动

- `tensorflow/python/ops/numpy_ops/np_math_ops.py`: branch on `x1 > x2` and anchor each branch on its own argument:

  ```python
  where(x1 > x2, x1 + log1p(exp(x2 - x1)),
                 x2 + log1p(exp(x1 - x2)))
  ```

  Both forms are the same `max + log1p(exp(-|x1 - x2|))` with a different anchor, so the forward values are unchanged, while each argument now owns its own differentiable expression and receives `0.5`. Both exponents are `<= 0`, so neither branch can overflow — this is why the obvious `(x1 + x2 + |delta|) / 2` or `x1 + softplus(x2 - x1)` reformulations were rejected: they overflow to `inf` for large-magnitude inputs of opposite sign.
- `tensorflow/python/ops/numpy_ops/np_math_ops_test.py`: add `testLogaddexpEqualInputsGradient`.

## 验证

- [x] `np_math_ops_test.py -k Logaddexp`: the new test fails before the change (gradients `1.0` / `0.0`) and passes after (`0.5` / `0.5`); the pre-existing `testLogaddexp` and `testLogaddexpNonFloatInputs` keep passing.
- [x] Full `MathTest`: 12 failures on the unpatched baseline, 11 after the change — no new failures. The remainder are pre-existing mismatches between the master test file and the runtime used locally.
- [x] Forward values are bit-for-bit identical for float16/float32/float64; max relative error against a `np.longdouble` reference is unchanged (3.7269770e-02 / 2.0656475e-05 / 5.3201161e-15), including `inf`, opposite-sign extremes and broadcast shapes.

## 影响面

No forward value change. Gradients at (and near) `x1 == x2` change from `(1.0, 0.0)` to the analytically correct `(0.5, 0.5)`, and second derivatives from `(0.0, 0.0)` to `(0.25, 0.25)`.

### CODE REVIEW

File: tensorflow/python/ops/numpy_ops/np_math_ops.py [hunk 1]
```diff
@@ -536,12 +536,21 @@ def f(x1, x2):
...
     return np_array_ops.where(
         isnan(delta),
         x1 + x2,  # NaNs or infinities of the same sign.
-        amax + log1p(exp(-abs(delta))),
+        np_array_ops.where(
+            x1 > x2,
+            x1 + log1p(exp(x2 - x1)),
+            x2 + log1p(exp(x1 - x2)),
+        ),
     )
```
[the remark this answers is not in the document]
Comment (Phoebus-Liu) [reply tensorflow/python/ops/numpy_ops/np_math_ops.py:553]: Thank you for the review, and for the suggestion to avoid the redundant `exp`/`log1p` evaluations.

I've applied it: the result is now computed once from `max_val`/`min_val` (selected through `where`) instead of materializing both `x1 > x2` branches. I verified that the forward values are bit-for-bit identical to the previous two-branch form and to the upstream `maximum` form, while halving the number of transcendental evaluations.

The gradient fix is unchanged: branching on `x1 > x2` keeps each argument's own differentiable expression, so both gradients are 0.5 at `x1 == x2` (previously 1.0 and 0.0 via `maximum`). Verified locally: `MathTest.testLogaddexp*`, the gradient (0.5/0.5), and the second derivative (0.25 at equality) all pass, and `inf`/`nan` behaviour is unchanged.

**Review by dmiltr3 — APPROVED** [review]
LGTM. Internal presubmit tests passed cleanly.

**Comment by Phoebus-Liu** [comment]
Thank you, @dmiltr3, for the review and approval! Glad the logaddexp gradient fix and the simplified `exp`/`log1p` evaluation passed the internal presubmit cleanly. Looking forward to the merge.

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Hi @Phoebus-Liu,

Thank you for your contribution and for implementing the fix for `tf.experimental.numpy.logaddexp`! The implementation and test cases have been reviewed and approved.

However, recent changes were merged to the upstream repository in `tensorflow/python/ops/numpy_ops/` (updating `np_math_ops.py` and `np_math_ops_test.py`) after this PR branch was created. Because the target files were modified upstream, the automated integration and presubmit testing pipeline encountered conflicts with out-of-date files when attempting to merge this change.

Could you please rebase your branch on the latest `master` (or merge `master` into your branch) and push the updated branch?

```bash
git fetch upstream
git checkout fix/issue-127232
git rebase upstream/master
git push --force-with-lease origin fix/issue-127232
```

Once the branch is synced with the latest upstream `master`, the automated synchronization and presubmit testing pipeline will re-evaluate the clean diff and proceed with merging.

Thanks again!

**Comment by Phoebus-Liu** [comment]
Hi @dmiltr3,

Thanks for the review and for confirming that the implementation and test cases are approved. I've rebased the branch onto the latest `master` (9edd58f8) and squashed everything into a single commit (9373389c), so the internal diff should no longer be stale.

No functional changes were made — the only updates are the rebase itself and flattening the history into one commit. Please let me know if anything else is needed on my side.

Thanks again for your time!
